### ISSUE TITLE
Generated ODS document without Table causes NPE

### ISSUE THREAD

**Issue by thadguidry** [OpenRefine/OpenRefine#2465]
**Describe the bug**
An ODS document without a `office:spreadsheet` element causes an NPE

**To Reproduce**
1. Download [NoData_NoSpreadsheet.ods.zip](https://github.com/OpenRefine/OpenRefine/files/4371589/NoData_NoSpreadsheet.ods.zip)
1. Rename to NoData_NoSpreadsheet.ods
1. Import the generated file NoData_NoSpreadsheet.ods  (just rename attached ODS package file)
1. OpenRefine recognizes as a valid ODF file (incorrectly) and attempts to parse as ODS format.
1. See error

**Current Results**
Fails silently and presents the user with a blank preview pane and no indication of error.

The server console/log shows the following stack trace:
```log
14:39:30.626 [                   refine] POST /command/core/importing-controller (3ms)
Mar 23, 2020 2:39:30 PM org.odftoolkit.odfdom.dom.OdfSchemaDocument getTables
SEVERE: null
java.lang.NullPointerException
        at org.odftoolkit.odfdom.dom.OdfSchemaDocument.fillTableList(OdfSchemaDocument.java:340)
        at org.odftoolkit.odfdom.dom.OdfSchemaDocument.getTables(OdfSchemaDocument.java:324)
        at org.odftoolkit.odfdom.doc.OdfDocument.getTableList(OdfDocument.java:692)
        at com.google.refine.importers.OdsImporter.parseOneFile(OdsImporter.java:142)
        at com.google.refine.importers.ImportingParserBase.parseOneFile(ImportingParserBase.java:118)
        at com.google.refine.importers.ImportingParserBase.parse(ImportingParserBase.java:89)
        at com.google.refine.importing.ImportingUtilities.previewParse(ImportingUtilities.java:961)
        at com.google.refine.importing.DefaultImportingController.doUpdateFormatAndOptions(DefaultImportingController.java:174)
        at com.google.refine.importing.DefaultImportingController.doPost(DefaultImportingController.java:93)
        at com.google.refine.commands.importing.ImportingControllerCommand.doPost(ImportingControllerCommand.java:68)
        at com.google.refine.RefineServlet.service(RefineServlet.java:190)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:820)
```
**Expected behavior**
OpenRefine should show an error dialog saying that the file is not a valid ODS document and include a link to the online ODF validator - https://odfvalidator.org/.

**Desktop:**
 - OS: Windows 10
 - Browser Version: Firefox
 - JRE or JDK Version: JDK8

**OpenRefine:**
 - Version OpenRefine 3.3

**Additional context**
The provided generated ODS file is actually NOT VALID, and failed parsing with [online hosted ODF Validator instance](https://odfvalidator.org/)
```
Result for NoData_NoSpreadsheet.ods

The document is NOT conformant ODF1.2!
Details:
NoData_NoSpreadsheet.ods: Info: ODF version of root document: 1.2
NoData_NoSpreadsheet.ods/mimetype: Error: The file 'mimetype' shall not be compressed in the ODF package 'NoData_NoSpreadsheet.ods'!
NoData_NoSpreadsheet.ods/mimetype: Error: The file 'mimetype' is not the first file in the ODF package 'NoData_NoSpreadsheet.ods'!
internal:/schema/odf1.2/OpenDocument-v1.2-os-manifest-schema.rng: Info: parsed.
NoData_NoSpreadsheet.ods/META-INF/manifest.xml: Info: no errors, no warnings
NoData_NoSpreadsheet.ods/mimetype: Info: 2 errors, no warnings
NoData_NoSpreadsheet.ods: Info: Media Type: application/vnd.oasis.opendocument.spreadsheet
internal:/schema/odf1.2/OpenDocument-v1.2-os-schema.rng: Info: parsed.
NoData_NoSpreadsheet.ods/meta.xml: Info: Generator: LibreOffice/6.0.3.2$Windows_X86_64 LibreOffice_project/8f48d515416608e3a835360314dac7e47fd0b821
NoData_NoSpreadsheet.ods/meta.xml: Info: no errors, no warnings
NoData_NoSpreadsheet.ods/settings.xml: Info: no errors, no warnings
NoData_NoSpreadsheet.ods/styles.xml: Info: no errors, no warnings
NoData_NoSpreadsheet.ods/content.xml[23,16]: Error: uncompleted content model. expecting: <chart>,<database>,<drawing>,<image>,<presentation>,<spreadsheet>,<text>
</office:body> ----^ NoData_NoSpreadsheet.ods/content.xml: Info: 1 errors, no warnings
internal:/schema/odf1.2/OpenDocument-v1.2-os-dsig-schema.rng: Info: parsed.
NoData_NoSpreadsheet.ods: Info: 3 errors, no warnings
```

**Comment by tfmorris** [comment]
I was making the instructions more explicit with the goal of tagging this as a Good First Issue since it's a simple NPE, when I came across the validation suggestion/requirement. Is that really necessary? It seems like checking for null and giving a simple error message would be a big improvement over the current behavior.

**Comment by thadguidry** [comment]
@tfmorris sure, in the section **Expected behavior** I gave some error message suggestions.

"not valid ODS file"
"No Tables found. Please validate and verify ODS file format"

whatever you think is useful.

**Comment by tfmorris** [comment]
OK, thanks. I've updated the expected behavior. 
@thadguidry When you say "generated," what was it generated by? (Trying to get a sense of priority ie how many files like this are in the wild)

**Comment by thadguidry** [comment]
@tfmorris low priority.  Only came across 2 files like this in the past 8 years :-)

**Comment by shriyasankhyan** [comment]
Can you assign me this issue  @wetneb ? I want to try to work on it.

**Comment by ComgLq24** [comment]
@thadguidry  Hello, I and my teammate @Unnamed-1408 was wondering whether we could give this issue a try. We have some questions.

1. If just to solve the NPE exception, we can just simply use the code(OdfSchemaDocument.class, getTables() method) below to judge, because is the `contentRoot` becomes `null` that causes NPE

```java
OfficeBodyElement officeBody = (OfficeBodyElement)OdfElement.findFirstChildNode(OfficeBodyElement.class,this.getContentDom().getRootElement());
OdfElement contentRoot = OdfElement.findFirstChildNode(OdfElement.class, officeBody);
tableList = this.fillTableList(contentRoot, (List)tableList);
```

However, if you need to check the ODS format, it needs to do more works. And I wonder if we just make sure it don't throw NPE for this example or make sure the file is well-formed.


2. It seems like we could judge loading is successful or not by checking the size of the tables in `List <OdfTable> tables = odfDoc.getTableList();`. However, there would still be NPE prompt on terminal. Whether it will be OK if there is NPE log on terminal but a loading-error prompt on client-end

**Comment by Unnamed-1408** [comment]
@tfmorris @wetneb @thadguidry Hello, can you answer the above questions, thanks a lot!

**Comment by thadguidry** [comment]
@ComgLq24 I listed in the original comment of what is the **Expected behavior**.  We have many examples in our code where we show a popup dialog during import.  Take a look at other importers to get a sense of that.  The biggest as mentioned above in the Current results is 
>Fails silently and presents the user with a blank preview pane and no indication of error.

Having an NPE in the log is acceptable, as long as on the frontend we have notified the user (and not failed silently) with a popup dialog about the invalid ODS file as I mentioned.
